tests/test_connection.py

```diff
@@ -0,0 +1,189 @@
+
+
+class ConnectionTestCase(unittest.TestCase):
+
+    def setUp(self):
+        self.connection = Connection()
+
+    def test_init(self):
+        pass
+
+    def test_build_params(self):
+        self.connection.api_key = 'test'
+        empty = {'data': {'a': 1}}
+        params = self.connection.build_params(empty)
+        self.assertNotEqual(params, empty)
+
+        self.assertTrue('timeout' in params)
+        self.assertEqual(params['timeout'], self.connection.timeout)
+
+        self.assertTrue('headers' in params)
+
+        self.assertTrue('Authorization' in params['headers'])
+        self.assertEqual(params['headers']['Authorization'], 'ApiKey {0}'.format(self.connection.api_key))
+
+        self.assertTrue('content-type' in params['headers'])
+        self.assertEqual(params['headers']['content-type'], self.connection.CONTENT_TYPE)
+
+        self.assertEqual(params['data'], '{"a": 1}')
+
+    def test_build_url(self):
+        result = urljoin(self.connection.host, 'test/')
+        self.assertEqual(self.connection.build_url('test'), result)
+        self.assertEqual(self.connection.build_url('/test'), result)
+        self.assertEqual(self.connection.build_url('/test/'), result)
+        self.assertEqual(self.connection.build_url(result), result)
+
+        with self.assertRaises(SyncanoValueError):
+            self.connection.build_url(True)
+
+    @mock.patch('syncano.connection.Connection.authenticate')
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_request_authentication(self, make_request_mock, authenticate_mock):
+        self.assertFalse(make_request_mock.called)
+        self.assertFalse(authenticate_mock.called)
+        self.connection.request('POST', 'test')
+        self.assertTrue(make_request_mock.called)
+        self.assertTrue(authenticate_mock.called)
+
+    @mock.patch('syncano.connection.Connection.authenticate')
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_request_serialization(self, make_request_mock, authenticate_mock):
+        self.assertFalse(make_request_mock.called)
+        self.assertFalse(authenticate_mock.called)
+
+        make_request_mock.return_value = {'a': 1}
+        content = self.connection.request('POST', 'test')
+
+        self.assertTrue(make_request_mock.called)
+        self.assertTrue(authenticate_mock.called)
+        self.assertEqual(content, make_request_mock.return_value)
+
+        content = self.connection.request('POST', 'test', result_class=mock.MagicMock)
+        self.assertIsInstance(content, mock.MagicMock)
+
+        make_request_mock.return_value = {
+            'objects': [],
+            'next': None,
+            'prev': None,
+        }
+        content = self.connection.request('POST', 'test')
+        self.assertIsInstance(content, self.connection.RESULT_SET_CLASS)
+
+    @mock.patch('requests.Session.post')
+    def test_make_request(self, post_mock):
+        response_mock = mock.MagicMock(
```

Comment:
I wonder if there are __ (dunder) methods used somewhere or if not, regular mock.Mock() would run just as well, being more lightweight.